server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This service provides access to this node's view of the master history, as well as access to other nodes' view of master stability.
+ */
+public class MasterHistoryService {
+    private final TransportService transportService;
+    private final MasterHistory localMasterHistory;
+    private final ClusterService clusterService;
+
+    /*
+     * This is a map of a node to the view of master history it has. This is populated asynchronously and is not guaranteed to have an
+     * entry for every node.
+     */
+    private final Map<DiscoveryNode, List<DiscoveryNode>> nodeToHistoryMap = new HashMap<>();
+    private static final Logger logger = LogManager.getLogger(MasterHistoryService.class);
+
+    public MasterHistoryService(
+        TransportService transportService,
+        Coordinator coordinator,
+        ThreadPool threadPool,
+        ClusterService clusterService
+    ) {
+        this.transportService = transportService;
+        this.localMasterHistory = new MasterHistory(threadPool, clusterService);
+        this.clusterService = clusterService;
+    }
+
+    /**
+     * This returns the MasterHistory as seen from this node. The returned MasterHistory will be automatically updated whenever the
+     * ClusterState on this node is updated with new information about the master.
+     * @return The MasterHistory from this node's point of view. This MasterHistory object will be updated whenever the ClusterState changes
+     */
+    public MasterHistory getLocalMasterHistory() {
+        return localMasterHistory;
+    }
+
+    /**
+     * This method returns a static view of the MasterHistory on the node given. This MasterHistory is static in that it will not be
+     * updated even if the ClusterState is updated on this node or the remote node. The history is retrieved asynchronously, and only if
+     * requestRemoteMasterHistory has been called for this node. If anything has gone wrong fetching it, or if it has not been fetched in
+     * time for the current request, the returned value will be null.
+     * @param node The node whose MasterHistory we want
+     * @return The MasterHistory from the remote node's point of view. This MasterHistory object will not be updated with future changes
+     */
+    @Nullable
+    public List<DiscoveryNode> getRemoteMasterHistory(DiscoveryNode node) {
+        synchronized (nodeToHistoryMap) {
+            return nodeToHistoryMap.get(node);
+        }
+    }
+
+    /**
+     * This method attempts to fetch the master history from the requested node. If we are able to successfully fetch it, it will be
+     * available in a later call to getRemoteMasterHistory. This method only fetches the remote master history once, and it is never
+     * updated unless this method is called again.
+     * @param node The node whose view of the master history we want to fetch
+     */
+    public void requestRemoteMasterHistory(DiscoveryNode node) {
```

Comment:
This method signature doesn't indicate this is an expensive call. Shall we have the ActionListener in its signature?

Reply:
Not sure I follow. What would we do with the ActionListener? And how do you define expensive? It returns immediately (doesn't block for the remote call). Do you mean expensive b/c it's making a remote call in the background?

Reply:
>And how do you define expensive?

A thing you wouldn't want to be called in a `for` loop. 

>Do you mean expensive b/c it's making a remote call in the background?

Yes

>What would we do with the ActionListener?

It would process the `response`. ie. 
```
nodeToHistoryMap.put(node, response.getMasterHistory());
```

Reply:
I'm confused -- the current ActionListener is doing exactly what your proposed new ActionListener would do, right?

Reply:
Naming suggestion that would help clarify perhaps?

```suggestion
    public void refreshRemoteMasterHistory(DiscoveryNode node) {
```